src/nvim/lua/xdiff.c

```diff
@@ -0,0 +1,323 @@
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+
+#include <stdlib.h>
+#include <stdio.h>
+#include <errno.h>
+#include <string.h>
+
+#include "nvim/vim.h"
+#include "nvim/xdiff/xdiff.h"
+#include "nvim/lua/xdiff.h"
+#include "nvim/lua/converter.h"
+#include "nvim/lua/executor.h"
+#include "nvim/api/private/helpers.h"
+
+typedef enum {
+  kNluaXdiffModeStr =  0,
+  kNluaXdiffModeCb,
+  kNluaXdiffModeLines,
+} NluaXdiffMode;
+
+typedef struct {
+  lua_State *lstate;
+  Error     *err;
+} hunkpriv_t;
+
+#ifdef INCLUDE_GENERATED_DECLARATIONS
+# include "lua/xdiff.c.generated.h"
+#endif
+
+static int write_string(void *priv, mmbuffer_t *mb, int nbuf)
+{
+  luaL_Buffer *buf = (luaL_Buffer *)priv;
+  for (int i = 0; i < nbuf; i++) {
+    const long size = mb[i].size;
+    for (long total = 0; total < size; total += LUAL_BUFFERSIZE) {
+      const int tocopy = MIN((int)(size - total), LUAL_BUFFERSIZE);
+      char *p = luaL_prepbuffer(buf);
+      if (!p) {
+        return -1;
+      }
+      memcpy(p, mb[i].ptr + total, (unsigned)tocopy);
+      luaL_addsize(buf, (unsigned)tocopy);
+    }
+  }
+  return 0;
+}
+
+// hunk_func callback used when opts.hunk_lines = true
+static int hunk_lines_func(long start_a, long count_a,
+                           long start_b, long count_b, void *cb_data)
+{
+  // Mimic extra offsets done by xdiff, see:
+  // src/nvim/xdiff/xemit.c:284
+  // src/nvim/xdiff/xutils.c:(356,368)
+  if (count_a > 0) {
```

Comment:
this adjustment is only done for `hunk_lines` and not `hunk_func`. we probably want the same values for both.

Reply:
I'm not sure tbh, it's not clear to me why this adjustment is done at such a high level in xdiff. Given how `hunk_func` is a low level callback, I think we should do a little as possible with it and just call it, otherwise we should rename `hunk_func` to something else to indicate it isn't the same `hunk_func` in xdiff. `hunk_lines` is something we've introduced and is a bit more high level, so it seems ok to make the adjustment there.

Reply:
Both `hunk_func` and `hunk_lines` is something _we_ introduce as mappings to the underlying xdiff interface. Whether the function or the table form is more appropriate to a _user_ of the API is unrelated to the point of the adjustment, so doing it in one entry point and not the other is arbitrary and confusing as an interface. This has no practical effect other than a completely unnecessary papercut whenever a user switches from one interface to the other.

Reply:
Ok, I'll make the adjustment in `hunk_func` too. Are you ok if I also rename `hunk_func` to `on_hunk` to indicate it isn't a 1-1 mapping, and to be more consistent with the naming scheme we use for other callbacks: `on_win`, `on_line`, etc?

Reply:
and `hunk_lines_func` to `on_lines`? (I'd like to have symmetry there as far as at all appropriate.)

Reply:
`hunk_lines` isn't a callback though. It just says to return a table instead of a string in unified format.

Reply:
That sounds good. :+1: